torch/_refs/__init__.py

```diff
@@ -0,0 +1,250 @@
+import torch
+from torch._C import _add_docstr  # type: ignore[attr-defined]
+
+import torch._prims as prims
+from functools import reduce
+
+from numbers import Number
+from typing import Sequence, Optional
+
+all = [
+  'add',
+]
+
+Tensor = torch.Tensor
+
+# Returns the Python type (AKA "type kind") corresponding to the torch dtype.
+# TODO: refactor into utils
+def _dtype_to_kind(dtype):
+  bools = (torch.bool,)
+  ints = (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64)
+  floats = (torch.float16, torch.bfloat16, torch.float32, torch.float64)
+  complexes = (torch.complex64, torch.complex128)
+
+  if dtype in bools:
+    return bool
+  if dtype in ints:
+    return int
+  if dtype in floats:
+    return float
+  if dtype in complexes:
+    return complex
+
+  raise ValueError("Invalid dtype!")
+
+# Returns the higher of two torch datatypes a and b or, if the two
+#   are not ordered relative to each other, the next
+#   higher datatype
+def _higher_dtype(a: torch.dtype, b: torch.dtype):
+  '''
+  Computes the "lowest" datatype that is weakly
+  "higher" than both a and b.
+  '''
+
+  # Type checking
+  assert isinstance(a, torch.dtype)
+  assert isinstance(b, torch.dtype)
+
+  if a is b:
+    return a
+
+  ordered_datatypes = (
+    (torch.bool,),
+    (torch.uint8, torch.int8),
+    (torch.int16,),
+    (torch.int32,),
+    (torch.int64,),
+    (torch.float16, torch.bfloat16),
+    (torch.float32,),
+    (torch.float64,),
+    (torch.complex64,),
+    (torch.complex128,)
+  )
+
+  for idx, dtypes in enumerate(ordered_datatypes):
+    if a in dtypes and b in dtypes:
+      return ordered_datatypes[idx + 1][0]
+    if a in dtypes:
+      return b
+    if b in dtypes:
+      return a
+
+def _higher_type_kind(a: Optional[type], b: Optional[type]):
+  ordered_type_kinds = (
+    bool, int, float, complex
+  )
+
+  assert a is None or a in ordered_type_kinds, a
+  assert b is None or b in ordered_type_kinds, b
+
+  if a is b:
+      return a
+
+  if a is None:
+    return b
+
+  if b is None:
+    return a
+
+  for kind in ordered_type_kinds:
+    if a is kind:
+      return b
+    if b is kind:
+      return a
+
+  raise ValueError("Received unknown types!")
```

Comment:
These things area already in `_prims` and here we're importing `_prims`. Can't we just import them?